args/core/scheduling/scheduler.hpp

```diff
@@ -0,0 +1,402 @@
+	{
+	private:
+		struct thread_error
+		{
+			std::string message;
+			std::thread::id threadId;
+		};
+
+		ProcessChain m_localChain;
+		async::readonly_rw_spinlock m_processChainsLock;
+		sparse_map<id_type, ProcessChain> m_processChains;
+		async::readonly_rw_spinlock m_errorsLock;
+		std::vector<thread_error> m_errors;
+
+		async::readonly_rw_spinlock m_exitsLock;
+		std::vector<std::thread::id> m_exits;
+
+		std::atomic_bool m_requestSync;
+		async::ring_sync_lock m_syncLock;
+
+		std::atomic<float> m_timeScale = 1.f;
+
+		inline static async::readonly_rw_spinlock m_threadsLock;
+		inline static sparse_map<std::thread::id, std::unique_ptr<std::thread>> m_threads;
+		inline static const uint m_maxThreadCount = std::thread::hardware_concurrency() == 0 ? std::numeric_limits<uint>::max() : std::thread::hardware_concurrency();
+		inline static async::readonly_rw_spinlock m_availabilityLock;
+		inline static uint m_availableThreads = m_maxThreadCount - 2; // subtract OS and this_thread.
+
+	public:
+		Scheduler()
+		{
+			addChain("Update");
+		}
+
+		~Scheduler()
+		{
+			for (auto& processChain : m_processChains)
+				processChain.exit();
+
+			for (auto& thread : m_threads)
+				if (thread->joinable())
+					thread->join();
+		}
+
+		/**@brief Set global time scale.
+		 */
+		void setTimeScale(float scale)
+		{
+			m_timeScale.store(scale, std::memory_order_relaxed);
+		}
+
+		/**@brief Get the global time scale.
+		 */
+		float getTimeScale()
+		{
+			return m_timeScale.load(std::memory_order_relaxed);
+		}
+
+		/**@brief Run main program loop, also starts all process-chains in their own threads.
+		 */
+		void run()
+		{
+			{
+				async::readonly_guard guard(m_processChainsLock);
+				for (ProcessChain& chain : m_processChains)
+					chain.run();
+			}
+
+			while (true) //TODO(glyn leine): Check for engine exit flag, needs engine event system/event bus.
+			{
+				{
+					async::readwrite_guard guard(m_exitsLock);
+
+					if (m_exits.size())
+					{
+						for (auto& id : m_exits)
+						{
+							destroyThread(id);
+						}
+
+						m_exits.clear();
+					}
+				}
+
+				{
+					async::readwrite_guard guard(m_errorsLock);
+
+					if (m_errors.size())
+					{
+						for (thread_error& error : m_errors)
+						{
+							std::cout << error.message << std::endl;
+							destroyThread(error.threadId);
+						}
+
+						m_errors.clear();
+
+						throw std::logic_error("");
+					}
+				}
+
+				{
+					std::vector<id_type> toRemove;
+					
+					{
+						async::readonly_multiguard rmguard(m_processChainsLock, m_threadsLock);
+						for (auto& chain : m_processChains)
+							if (!m_threads.contains(chain.threadId()))
+								toRemove.push_back(chain.id());
+					}
+
+					async::readwrite_guard wguard(m_processChainsLock);
+					for (id_type& id : toRemove)
+						m_processChains.erase(id);
+				}
+
+				if (m_localChain.id())
+					m_localChain.runInCurrentThread();
+
+				if (syncRequested())
+					waitForProcessSync();
+			}
+		}
```

Comment:
header only is neat and all, but I would move this to a .cpp for readability

Reply:
well you asked before on a different class if ARGS_API was rly necessary, and since it's not rly necessary here i thought i'd avoid using ARGS_API